pyfar/dsp/dsp.py

```diff
@@ -312,6 +312,263 @@ def spectrogram(signal, window='hann', window_length=1024,
+        and 1 otherwise.
+    freq_range: tuple
+        Two element vector specifying upper and lower frequency bounds
+        for normalization.
+
+    Returns
+    --------
+    normalized_signal: Signal
+        The normalized signal
+    values: numpy array
+        If return_values=True returns values, the values of all channels
+        before normalization.
+    """
+
+    # check input
+    if not isinstance(signal, pyfar.Signal):
+        raise TypeError('Input data has to be of type: Signal.')
+
+    # set default values
+    if value is None:
+        value = 0 if normalize == 'log_magnitude' else 1
+    if freq_range is None:
+        freq_range = (0, signal.frequencies[-1])
+
+    # copy and transform data to the desired domain
+    if normalize == 'time':
+        input_data = np.abs(signal.time.copy())
+    elif normalize == 'magnitude':
+        input_data = np.abs(signal.freq.copy())
+    elif normalize == 'log_magnitude':
+        input_data = 20 * np.log10(signal.freq.copy())
+    else:
+        raise ValueError(
+                    "normalize must be 'time', 'magnitude' or 'log_magnitude'")
+
+    # get bounds for normalization
+    if normalize == 'time':
+        lim = (0, signal.n_samples)
+
+    else:
+        lim = signal.find_nearest_frequency(freq_range)
+        if signal.n_samples % 2:
+            lim[0] = np.max([lim[0], 1])
+        else:
+            lim = np.clip(lim, 1, signal.n_bins-1)
+
+    # get values for normalization
+    if normalize_to == 'max':
+        values = np.max(input_data[..., lim[0]:lim[1]], axis=-1,
+                        keepdims=True)
+    elif normalize_to == 'mean':
+        values = np.mean(input_data[..., lim[0]:lim[1]], axis=-1,
+                         keepdims=True)
+    elif normalize_to == 'rms':
+        values = np.sqrt(np.mean(input_data[..., lim[0]:lim[1]]**2,
+                         axis=-1, keepdims=True))
+    else:
+        raise ValueError("normalize_to must be 'max', 'mean' or 'rms'")
+
+    # manipulate values
+    if channel_handling == 'each':
+        pass
+    elif channel_handling == 'max':
+        values = np.max(values)
+    elif channel_handling == 'min':
+        values = np.min(values)
+    elif channel_handling == 'mean':
+        values = np.mean(values)
+    else:
+        raise ValueError(
+                    "channel_handling must be 'each', 'max', 'min' or 'mean'")
+
+    # de-logarthimize value
+    if normalize == 'log_magnitude':
+        value = 10**(value/20)
+
+    # replace input with normalized_input
+    normalized_signal = signal.copy()
+    if normalize == 'time':
+        normalized_signal.time = signal.time.copy() / values * value
+    else:
+        normalized_signal.freq = signal.freq.copy() / values * value
+
+    if return_values:
+        return normalized_signal, values
+    else:
+        return normalized_signal
+
+
+def average(signal, average_mode='time', phase_copy=None,
```

Comment:
I think we should ad an `axis` parameter that allows us to average over all or only a specific axis.